tensorflow/contrib/kafka/kernels/kafka_reader_ops.cc

```diff
@@ -0,0 +1,190 @@
+    if (parts.size() < 1) {
+      return errors::InvalidArgument("Invalid parameters: ", current_work());
+    }
+    string topic = parts[0];
+    int32 partition = 0;
+    if (parts.size() > 1) {
+      if (!strings::safe_strto32(parts[1], &partition)) {
+        return errors::InvalidArgument("Invalid parameters: ", current_work());
+      }
+    }
+    int64 offset = 0;
+    if (parts.size() > 2) {
+      if (!strings::safe_strto64(parts[2], &offset)) {
+        return errors::InvalidArgument("Invalid parameters: ", current_work());
+      }
+    }
+
+    topic_partition_.reset(
+        RdKafka::TopicPartition::create(topic, partition, offset));
+
+    offset_ = topic_partition_->offset();
+    limit_ = -1;
+    if (parts.size() > 3) {
+      if (!strings::safe_strto64(parts[3], &limit_)) {
+        return errors::InvalidArgument("Invalid parameters: ", current_work());
+      }
+    }
+
+    std::unique_ptr<RdKafka::Conf> conf(
+        RdKafka::Conf::create(RdKafka::Conf::CONF_GLOBAL));
+    std::unique_ptr<RdKafka::Conf> topic_conf(
+        RdKafka::Conf::create(RdKafka::Conf::CONF_TOPIC));
+
+    std::string errstr;
+
+    RdKafka::Conf::ConfResult result =
+        conf->set("default_topic_conf", topic_conf.get(), errstr);
+    if (result != RdKafka::Conf::CONF_OK) {
+      return errors::Internal("Failed to set default_topic_conf:", errstr);
+    }
+
+    result = conf->set("bootstrap.servers", servers_, errstr);
+    if (result != RdKafka::Conf::CONF_OK) {
+      return errors::Internal("Failed to set bootstrap.servers ", servers_, ":",
+                              errstr);
+    }
+    result = conf->set("group.id", group_, errstr);
+    if (result != RdKafka::Conf::CONF_OK) {
+      return errors::Internal("Failed to set group.id ", group_, ":", errstr);
+    }
+
+    consumer_.reset(RdKafka::KafkaConsumer::create(conf.get(), errstr));
+    if (!consumer_.get()) {
+      return errors::Internal("Failed to create consumer:", errstr);
+    }
+
+    std::vector<RdKafka::TopicPartition*> partitions;
+    partitions.emplace_back(topic_partition_.get());
+    RdKafka::ErrorCode err = consumer_->assign(partitions);
+    if (err != RdKafka::ERR_NO_ERROR) {
+      return errors::Internal(
+          "Failed to assign partition [", topic_partition_->topic(), ", ",
+          topic_partition_->partition(), ", ", topic_partition_->offset(), "]:",
+          RdKafka::err2str(err));
+    }
+
+    return Status::OK();
+  }
+
+  Status OnWorkFinishedLocked() override {
+    consumer_->unassign();
+    consumer_->close();
+    consumer_.reset(nullptr);
+    return Status::OK();
+  }
+
+  Status ReadLocked(string* key, string* value, bool* produced,
+                    bool* at_end) override {
+    if (limit_ >= 0 &&
+        (topic_partition_->offset() >= limit_ || offset_ >= limit_)) {
+      *at_end = true;
+      return Status::OK();
+    }
+    while (true) {
+      std::unique_ptr<RdKafka::Message> message(consumer_->consume(1000));
```

Comment:
Not too familiar with `rdkafkacpp` but any reasons behind this hard-coded 1000? Is this a standard practice?

Reply:
Thanks @terrytangyuan. The `1000` is the timeout (in `ms`) for consume() call to return. The call will repeat in the loop. The value was used as in the demo code of librdkafka.

I have updated the PR so that an optional attribute `timeout` could be passed to control this value (instead of hard coded `1000`). Please take a look.